### PR TITLE
[WebDriver][BiDi] Implement the browsingContext.traverseHistory command

### PR DESCRIPTION
**Pull request by AnujPanta1**
#### 2ff46eecb236c2ddadd70ced3c8b4c042f2ffb0a
<pre>
[WebDriver][BiDi] Implement the browsingContext.traverseHistory command
<a href="https://bugs.webkit.org/show_bug.cgi?id=288334">https://bugs.webkit.org/show_bug.cgi?id=288334</a>
<a href="https://rdar.apple.com/problem/145987781">rdar://problem/145987781</a>

Reviewed by BJ Burg.

Adds traverseHistory support to navigate browsing context history by delta.
Includes bounds checking against the back/forward list to prevent invalid
navigation attempts beyond available history entries. Uses existing
WebBackForwardList to determine valid target indices and retrieve the
history item. The core navigation logic resides in WebAutomationSession.

* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
(WebKit::BidiBrowsingContextAgent::traverseHistory):
* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::goForwardInBrowsingContext):
(WebKit::WebAutomationSession::traverseHistoryInBrowsingContext):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:

Canonical link: <a href="https://commits.webkit.org/301579@main">https://commits.webkit.org/301579@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/053f178ee2156534720fc85bde0b77266f06bd23

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/126365 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/46016 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36897 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133312 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78098 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46687 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54564 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96197 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/129313 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37344 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113037 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76673 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/configuration (failure)") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36233 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76596 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/107155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31470 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135831 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53096 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40802 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104694 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53563 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109311 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104396 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26637 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49848 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28188 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/50489 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52999 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/58827 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52304 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55644 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54027 "Built successfully") | | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 1]
```diff
@@ -850,6 +851,45 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+    auto page = webPageProxyForHandle(handle);
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!page, WindowNotFound);
+
+    RefPtr mainFrame = page->mainFrame();
+    if (!mainFrame || !mainFrame->isMainFrame()) {
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InvalidParameter);
+        return;
+    }
+
+    if (delta <= 0) {
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp]: Shouldn't we support delta of -1?

File: Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json [hunk 2]
```diff
@@ -162,6 +162,17 @@
                 { "name": "url", "type": "string", "description": "The URL the browsing context navigated to." },
                 { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "optional": true, "description": "The navigation ID, or null if not applicable." }
             ]
+        },
+        {
+            "name": "traverseHistory",
+            "description": "Traverses the history of a given navigable by a delta.",
+            "spec": "https://w3c.github.io/webdriver-bidi/#command-browsingContext-traverseHistory",
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/browsing_context/traverse_history",
+            "async": true,
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json]: Nit: usually I put async last

File: Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json [hunk 3]
```diff
@@ -162,6 +162,17 @@
                 { "name": "url", "type": "string", "description": "The URL the browsing context navigated to." },
                 { "name": "navigation", "$ref": "BidiBrowsingContext.Navigation", "optional": true, "description": "The navigation ID, or null if not applicable." }
             ]
+        },
+        {
+            "name": "traverseHistory",
+            "description": "Traverses the history of a given navigable by a delta.",
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json]: Might want to mention that it errors out if the bounds check for `delta` fails.

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 4]
```diff
@@ -850,6 +851,45 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+
+    Ref backForwardList = page->backForwardList();
+
+    unsigned backCount = backForwardList->backListCount();
+    unsigned forwardCount = backForwardList->forwardListCount();
+
+    int currentIndex = static_cast<int>(backCount);
+    int targetIndex = currentIndex + delta;
+
+    if (targetIndex < 0 || targetIndex >= static_cast<int>(backCount + forwardCount + 1)) {
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp]: Please rewrite these error checks to use `ASYNC_FAIL_IF_PREDEFINED_ERROR_IF` and remove the if() indentation. Feel free to put the condition itself into a local variable to make the lines more readable if it helps.

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 5]
```diff
@@ -850,6 +851,45 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+    }
+
+    Ref backForwardList = page->backForwardList();
+
+    unsigned backCount = backForwardList->backListCount();
+    unsigned forwardCount = backForwardList->forwardListCount();
+
+    int currentIndex = static_cast<int>(backCount);
+    int targetIndex = currentIndex + delta;
+
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp]: Nit: extra newline

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 6]
```diff
@@ -850,6 +851,45 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+    if (delta <= 0) {
+        callback({ });
+        return;
+    }
+
+    Ref backForwardList = page->backForwardList();
+
+    unsigned backCount = backForwardList->backListCount();
+    unsigned forwardCount = backForwardList->forwardListCount();
+
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp]: Nit: extra newline

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 7]
```diff
@@ -850,6 +851,45 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+        return;
+    }
+
+    if (delta <= 0) {
+        callback({ });
+        return;
+    }
+
+    Ref backForwardList = page->backForwardList();
+
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp]: Nit: extra newline

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 8]
```diff
@@ -850,6 +851,45 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+    int targetIndex = currentIndex + delta;
+
+    if (targetIndex < 0 || targetIndex >= static_cast<int>(backCount + forwardCount + 1)) {
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+        return;
+    }
+
+    RefPtr targetItem = backForwardList->itemAtIndex(targetIndex);
+    if (!targetItem) {
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp]: Ditto re: macro use.

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 9]
```diff
@@ -850,6 +851,45 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+        return;
+    }
+
+    page->goToBackForwardItem(*targetItem);
+    waitForNavigationToCompleteOnPage(*page, defaultPageLoadStrategy, defaultPageLoadTimeout, WTFMove(callback));
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:888]: I am a little concerned that there might be some edge cases where `goToBackForwardItem()` doesn't actually cause a navigation, and therefore this will wait indefinitely. I think test failures will reveal whether this approach is deficient. Let's move forward for now since it's just for `bc.traverseHistory`.

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 10]
```diff
@@ -850,6 +851,45 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
     waitForNavigationToCompleteOnPage(*page, pageLoadStrategy, pageLoadTimeout, WTFMove(callback));
 }
 
+void WebAutomationSession::traverseHistoryInBrowsingContext(const Inspector::Protocol::Automation::BrowsingContextHandle& handle, int delta, CommandCallback<void>&& callback)
+{
+    auto page = webPageProxyForHandle(handle);
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!page, WindowNotFound);
+
+    RefPtr mainFrame = page->mainFrame();
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:868]: > If navigable is not a [top-level traversable](https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable), return [error](https://w3c.github.io/webdriver/#errors) with [error code](https://w3c.github.io/webdriver/#dfn-error-code) [invalid argument](https://w3c.github.io/webdriver/#dfn-invalid-argument).

I think this could be expressed more simply with `ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF`.

But as it is now, the `Automation.BrowsingContextHandle` can only represent a main frame/page, so this extra mainFrame check isn't necessary. This sort of check would be more appropriate with `BidiBrowsingContext.BrowsingContext` which can represent either. But in that case, shouldn't the logic should be in the Bidi agent? All things considered I'd prefer to put make WebAutomationSession smaller not larger :)

**Review by burg — CHANGES_REQUESTED** [review]

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp (hunk 1 above)
Comment (AnujPanta1) [reply Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp]: opp sorry this was supposed to be == 0 I think so we just do nothing

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 11]
```diff
@@ -855,6 +856,36 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+    if (!mainFrame || !mainFrame->isMainFrame()) {
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp]: I don't think we need this `isMainFrame` check, it's redundant.

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 12]
```diff
@@ -855,6 +856,36 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InvalidParameter);
+        return;
+    }
+
+    if (!delta) {
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:871]: Is it allowed to have delta of 0?

File: Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json [hunk 13]
```diff
@@ -162,6 +162,17 @@
...
             ]
+        },
+        {
+            "name": "traverseHistory",
+            "description": "Traverses the history of a given navigable by a delta. Errors out if the bounds check for delta fails.",
+            "spec": "https://w3c.github.io/webdriver-bidi/#command-browsingContext-traverseHistory",
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/browsing_context/traverse_history",
+            "parameters": [
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext", "description": "The identifier of the browsing context to traverse history." },
+                { "name": "delta", "type": "integer", "description": "The number of steps to traverse in history. Positive values go forward, negative values go backward." }
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json]: If we keep 0 as a valid value, then please note in the description that a value of 0 will always succeed and do nothing.

**Review by burg — APPROVED** [review]
This looks almost ready, just a few nits.

File: Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp [hunk 14]
```diff
@@ -855,6 +856,33 @@ void WebAutomationSession::goForwardInBrowsingContext(const Inspector::Protocol:
...
+    }
+
+    Ref backForwardList = page->backForwardList();
+    unsigned backCount = backForwardList->backListCount();
+    unsigned forwardCount = backForwardList->forwardListCount();
+    int currentIndex = static_cast<int>(backCount);
+    int targetIndex = currentIndex + delta;
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(targetIndex < 0, InvalidParameter);
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(targetIndex >= static_cast<int>(backCount + forwardCount + 1), InvalidParameter);
+    RefPtr targetItem = backForwardList->itemAtIndex(targetIndex);
```
Comment (burg) [inline Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:884]: Nit: I usually add an extra newline after ASYNC_FAIL_*

**Review by burg — APPROVED** [review]
